benchmark/utils/timer.hpp

```diff
@@ -0,0 +1,283 @@
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <chrono>
+#include <memory>
+
+
+#include <gflags/gflags.h>
+
+
+#ifdef HAS_CUDA
+
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+
+#include "cuda/base/device_guard.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+#ifdef HAS_HIP
+
+
+#include <hip/hip_runtime.h>
+
+
+#include "hip/base/device_guard.hip.hpp"
+
+
+#endif  // HAS_HIP
+
+
+DEFINE_bool(gpu_timer, false,
+            "use gpu timer based on event. It is valid only when "
+            "executor is cuda or hip");
+
+
+class Timer {
+public:
+    void tic()
+    {
+        assert(tic_called_ == false);
+        this->tic_impl();
+        tic_called_ = true;
+    }
+
+    std::int64_t toc()
+    {
+        assert(tic_called_ == true);
+        auto ns = this->toc_impl();
+        tic_called_ = false;
+        this->add_record(ns);
+        return ns;
+    }
+
+    std::int64_t get_total_time() const { return total_duration_ns_; }
+
+    std::int64_t get_num_repetitions() const { return duration_ns_.size(); }
+
+    double get_average_time() const
+    {
+        return static_cast<double>(this->get_total_time()) /
+               this->get_num_repetitions();
+    }
+
+    std::vector<std::int64_t> get_time_detail() const{ return duration_ns_; }
+
+    void clear()
+    {
+        duration_ns_.clear();
+        tic_called_ = false;
+        total_duration_ns_ = 0;
+    }
+
+    Timer() : tic_called_(false), total_duration_ns_(0) {}
+
+protected:
+    void add_record(std::int64_t ns)
+    {
+        // add the result;
+        duration_ns_.emplace_back(ns);
+        total_duration_ns_ += ns;
+    }
+
+    virtual void tic_impl() = 0;
+
+    virtual std::int64_t toc_impl() = 0;
+
+private:
+    std::vector<std::int64_t> duration_ns_;
+    bool tic_called_;
+    std::int64_t total_duration_ns_;
+};
+
+
+class CpuTimer : public Timer {
+public:
+    CpuTimer(std::shared_ptr<const gko::Executor> exec) : Timer(), exec_(exec)
+    {}
+
+protected:
+    void tic_impl() override
+    {
+        exec_->synchronize();
+        start_ = std::chrono::steady_clock::now();
+    }
+
+    std::int64_t toc_impl() override
+    {
+        exec_->synchronize();
+        auto stop = std::chrono::steady_clock::now();
+        auto duration_time =
+            std::chrono::duration_cast<std::chrono::nanoseconds>(stop - start_)
+                .count();
+        return static_cast<std::int64_t>(duration_time);
```

Comment:
I guess a cast is not required here anymore.
```suggestion
        return duration_time;
```